apps/emqx_cluster_link/src/emqx_cluster_link.erl

```diff
@@ -109,6 +112,44 @@ forward(Routes, Delivery) ->
 %%--------------------------------------------------------------------
 
 on_message_publish(
+    #message{topic = <<?ROUTE_TOPIC_PREFIX, _/binary>>} = Msg
+) ->
+    case handle_route_op_msg(Msg) of
+        ok ->
+            {stop, []};
+        error ->
+            %% Disconnect so that upstream agent starts anew
+            Headers = Msg#message.headers,
+            {stop, Msg#message{
+                headers = Headers#{
```

Comment:
BTW there's also `emqx_message:set_headers/2` (not sure though if that would be more readable).